### PR TITLE
Add difficulty calculation mod combinations

### PR DESCRIPTION
**Pull request by smoogipoo**
Allows the diffcalc server to compute the beatmap difficulty all mod combinations.

### CODE REVIEW

**Review by DrabWeb — DISMISSED** [review]

File: osu.Game.Rulesets.Osu/OsuRuleset.cs [hunk 1]
```diff
@@ -143,7 +141,47 @@ public override IEnumerable<Mod> GetModsFor(ModType type)
                         },
                         new OsuModTarget(),
                     };
-
+                case ModType.DifficultyCalculation:
+                    return new Mod[]
+                    {
+                        new MultiMod(),
```
Comment (peppy) [inline osu.Game.Rulesets.Osu/OsuRuleset.cs]: what is this?

**Comment by peppy** [comment]
This implementation seems a bit weird. Can we not do this by:

```csharp
var mods = GetModsFor(Increase | Reduction).Where(m => m.AffectsDifficulty /* if required */);

var combinations = new List<Mod>();

foreach (var m in mods)
    combinations.AddRange(createPermutations(m));
```

**Comment by smoogipoo** [comment]
I would need a good reason to move away from the currently accepted paradigm of using ModType to get the mods you're interested in.

Either we use ModType or we use properties in mods, not both. This is simplest.

File: osu.Game.Rulesets.Osu/OsuRuleset.cs (hunk 1 above)
Comment (smoogipoo) [reply osu.Game.Rulesets.Osu/OsuRuleset.cs]: This is supposed to be equivalent to no mod, alternative being we use nulls which sounds wrong, or using an `EmptyMod` type.

**Comment by peppy** [comment]
`ModType` wasn't intended to have arbitrary (not-displayed) types though.. it was purely for the display of mods at song select. The property may not be necessary (see comment in my proposal), was just covering that scenario.

i'm not sure where the idea of "accepted paradigm" comes from.. maybe i'm missing something? isn't this a completely new implementation?

File: osu.Game.Rulesets.Osu/OsuRuleset.cs (hunk 1 above)
Comment (peppy) [reply osu.Game.Rulesets.Osu/OsuRuleset.cs]: Definitely worth making a `NoModMod` to disambiguate

**Comment by smoogipoo** [comment]
The property is necessary, otherwise mod combinations that lead to "difficulty increase" such as FL and HD would be considered.

Besides that your example shows the easiest part of the process, which is retrieving the mods, but the recursive function is going to be painful to implement in a way that generically considers all rulesets (mania vs osu, as an example).

As for your argument that it wasn't intended to have not-displayed mods... Says who? It's an enum that retrieves some mods, nothing more than that.

**Comment by peppy** [comment]
My concern is that this decreases maintainability; it increases the number of places that need to be updated on any mod changes.

I can give the recursive function a shot if you'd like. I think it's possible.

**Comment by smoogipoo** [comment]
? How does it increase places to update if any mod changes? It's irrelevant unless diff calc changes.

**Comment by peppy** [comment]
It may be that the mods which should not be considered (FL/HD) should be decided by the difficulty calculator itself. Potentially those mods could be used in the future for difficulty calculation; I think it's best to leave this up to the point of calculation, not the ruleset returning mods.

**Comment by peppy** [comment]
Yes, but now you're impeding difficulty calculation specifics inside `Ruleset`. We should be avoiding these kinds of cross-dependencies where possible.

**Comment by smoogipoo** [comment]
What? No you're not. There would be no difference to diff calc unless you change the difficulty calculator in the first place. THATs the thing that impedes you - not having code that considers it in the first place.

And if you do write code to consider the mod you're making a conscious change to now calculate combinations including mod.

**Comment by peppy** [comment]
Hmm, do you have anything against the difficulty calculator being responsible for choosing which mods to consider / combinations to generate? I think it fits much better.

File: osu.Game.Rulesets.Mania/ManiaRuleset.cs [hunk 2]
```diff
@@ -96,6 +96,20 @@ public override IEnumerable<Mod> ConvertLegacyMods(LegacyMods mods)
                 yield return new ManiaModSuddenDeath();
         }
 
+        private static readonly Mod[] key_mods =
```
Comment (peppy) [inline osu.Game.Rulesets.Mania/ManiaRuleset.cs]: unused?

Comment (smoogipoo) [reply osu.Game.Rulesets.Mania/ManiaRuleset.cs]: Indeed.

**Review by peppy — APPROVED** [review]
much better now 🍡

**Pull request merged**